### PR TITLE
Support removing archived settings if cluster has read only blocks.

### PR DESCRIPTION
**Pull request by howardhuanghua**
Currently if we have follow cluster settings(we have set both persistent and transient before):
```
GET_cluster/settings?pretty
{
  "persistent" : {
    "cluster" : {
      "blocks" : {
        "read_only_allow_delete" : "true"
      }
    }
  },
  "transient" : {
    "cluster" : {
      "blocks" : {
        "read_only_allow_delete" : "true"
      }
    }
  }
}
``` 

Then we cannot remove them together:
```
PUT _cluster/settings?pretty
{
  "persistent" : {
    "cluster.blocks.read_only_allow_delete": null
  },
  "transient" : {
    "cluster.blocks.read_only_allow_delete": null
  }
}'

Response:
{
  "error" : {
    "root_cause" : [
      {
        "type" : "cluster_block_exception",
        "reason" : "blocked by: [FORBIDDEN/13/cluster read-only / allow delete (api)];"
      }
    ],
    "type" : "cluster_block_exception",
    "reason" : "blocked by: [FORBIDDEN/13/cluster read-only / allow delete (api)];"
  },
  "status" : 403
}

```

Even we can remove them one by one, but I think it's reasonable to allow user to remove them together. This PR supports both of the cases.

### CODE REVIEW

**Comment by rjernst** [comment]
@DaveCTurner Can you review please?

**Comment by DaveCTurner** [comment]
Does this address https://github.com/elastic/elasticsearch/issues/62920 @howardhuanghua? It's the right area of code to do so. If it does, could you add a test for that issue too? If it doesn't, would you do so?

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner, this is another seperated issue, it cannot fix https://github.com/elastic/elasticsearch/issues/62920 driectly.
Before fixing https://github.com/elastic/elasticsearch/issues/62920, I would like to fix this individual issue first.
I have also adjusted test case `testClusterUpdateSettingsWithBlocks` L323 in ClusterSettingsIT.java to support **removing persistent and transient cluster block settings together**. Without this fix, we could not set them together.

**Comment by DaveCTurner** [comment]
I would rather you addressed #62920 as well here. There is little value in us reviewing a PR which addresses a very minor issue and will be superseded straight away by a fix to a much more serious bug.

**Comment by howardhuanghua** [comment]
OK, I will update the PR to fix them together.

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner , I have updated the PR. Here are the related logics:

If cluster has read only blocks and archived settings:
1. Cannot remove read only blocks due to archived settings.
2. Cannot update any cluster settings due to read only blocks.
3. Can only remove archived settings.

That means we cannot do any changes before removing archived settings. And if cluster is read only, except removing archived settings, we could do nothing include removing read only blocks.

Now we also support removing transient and persistent cluster blocks together.

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner would you please help to check the updated logic again? Thanks.

**Comment by howardhuanghua** [comment]
Merged master.

**Comment by DaveCTurner** [comment]
@elasticmachine ok to test

**Review by DaveCTurner — COMMENTED** [review]
Permitting multiple settings in updates does open up quite a lot of complexity. For instance, with the suggested implementation here I think we would permit removing archived settings without removing the metadata block. You could even be adding a metadata block at the same time as removing those settings.

However, I don't think it's worth resolving the provided settings updates against the current cluster state here just for validation. Let's skip calling `globalBlockedException` only if all of the following are true:

- at least one of `cluster.blocks.read_only` or `cluster.blocks.read_only_allow_delete` is being cleared (set to `null` or `false`)
- neither `cluster.blocks.read_only` nor `cluster.blocks.read_only_allow_delete` is being set to `true`
- the only other settings in this update are archived ones being set to `null`.

I see that it's a bit of a pain to do this for both the transient and persistent settings at the same time, but could you find a way to do these checks without duplicating the entire loop and all its logic? Extract a method or two perhaps.

**Comment by DaveCTurner** [comment]
@elasticmachine update branch

**Comment by DaveCTurner** [comment]
@howardhuanghua you'll need to tick the box to let maintainers push code to your branch, or else look after merging latest `master` yourself. I've enabled CI on this branch anyway so it'll run the full test suite each time you push changes.

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner I have updated the latest master. And I will check your suggestions and remove duplicated loop. Thanks.

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner , I have updated the skip calling `globalBlockedException` logic based on your suggestion. I used a map to group by `transent` and `persistent` clear block and archived settings. One condition different is that we need to allow user to remove cluster blocks if cluster doesn't have archived settings.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Thanks @howardhuanghua, this looks closer. I left some more comments.

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner , I have updated the change based on your suggestions. Thank you.

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner , would you please help to review the changes again? Thanks a lot.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java:376]: I think here we should set these settings randomly, not just setting the one we care about to `true`.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java:311]: We can assert the persistent settings here too?

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java:314]: We can assert the persistent settings here too?

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java:333]: We could also set this to `null`.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/settings/ClusterSettingsIT.java:336]: We could also set this to `null`.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java:101]: This could just be a `Set<String>`, we don't care about the values.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java:106]: I don't think we should duplicate the parsing of boolean settings here. Suggest using `Metadata.SETTING_READ_ONLY_ALLOW_DELETE_SETTING.get(settings)` etc.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I left a few nits but this is looking good. I also pushed a branch with my suggested changes: https://github.com/TencentCloudES/elasticsearch/compare/remove_block...DaveCTurner:2021-01-26-remove_block-reviews

**Comment by howardhuanghua** [comment]
Hi @DaveCTurner , thanks for your patient help. I have reviewed and merged your suggested changes. Also successfuly run several times of `testRemoveArchiveSettingsWithBlocks` test case in my local env.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
Thanks @howardhuanghua, I'll get this merged shortly although probably not today.

**Pull request merged**
